### ISSUE TITLE
Orchestration research: insights from TRINITY + The Conductor (Sakana AI, Dec 2025)

### ISSUE THREAD

**Issue by jerryfane** [gitmoot/gitmoot#378]
## Why this issue

Two Sakana AI papers (Dec 2025) study **macro-level model fusion via coordination** —
composing a pool of black-box LLMs you *cannot* weight-merge into one stronger system. That is
exactly what `gitmoot orchestrate` does: a coordinator + a validated `delegations[]` DAG fanning
work to codex / claude / kimi and reconvening in a continuation job.

This issue is a **research reference, not a build commitment**. It records what transfers to
gitmoot's **prompt/contract-driven** coordinator, what explicitly does **not**, and the concrete
existing primitives each idea would layer on. Mine it later; nothing here is scheduled.

Conceptually builds on the orchestration primitives already shipped — #367, #368, #369, #370,
#371 — and relates to the orchestration epic #357.

## References / sources

| Paper | arXiv | Authors | One-line |
|---|---|---|---|
| **TRINITY: An Evolved LLM Coordinator** | [abs/2512.04695](https://arxiv.org/abs/2512.04695) · [pdf](https://arxiv.org/pdf/2512.04695) | Sakana AI | A *trained* <20K-param coordinator (frozen 0.6B SLM + ~10K linear head) routes 7 black-box LLMs over Thinker/Worker/Verifier roles, stops on Verifier ACCEPT; sep-CMA-ES on binary rewards; 86.2% LiveCodeBench SOTA. |
| **The Conductor: Learning to Orchestrate Agents in Natural Language** | [abs/2512.04388](https://arxiv.org/abs/2512.04388) · [pdf](https://arxiv.org/pdf/2512.04388) | Sakana AI | A *trained* 7B coordinator (GRPO) emits a natural-language workflow (subtasks + model ids + a per-step "access list"); trained on 960 problems, beats every worker incl. GPT-5 with ~3 calls avg. |

## The two papers at a glance

| | TRINITY | The Conductor |
|---|---|---|
| Coordinator | 0.6B SLM + ~10K-param head (<20K learnable) | 7B LLM |
| How trained | sep-CMA-ES (derivative-free), binary task reward | RL (GRPO), verifiable reward |
| Output per turn | which LLM + which role (Thinker/Worker/Verifier) | NL workflow: subtasks + model ids + access list |
| Pool | 7 black-box LLMs (GPT-5, Gemini-2.5-Pro, Claude-4-Sonnet, + open) | frontier + open workers |
| Stop signal | Verifier **ACCEPT** | last worker output scored vs ground truth |
| Headline | 86.2% LiveCodeBench (SOTA); beats every single model | beats every worker incl. GPT-5, fewer calls |
| **For gitmoot** | role contracts + verifier-gated stopping + complementary pool | per-edge visibility + difficulty-adaptive, fewer targeted calls |

## ⚠️ The framing caveat — read this first

Both *winning* coordinators are **trained**. Both papers explicitly tested the
**untrained-frontier-LLM-as-coordinator** baseline — **which is precisely gitmoot's regime** — and
it was **substantially worse** and over-relied on reputation priors:

| Setting | Coordinator | Result | Notes |
|---|---|---|---|
| TRINITY | trained <20K head | **70.44** | — |
| TRINITY | prompted frontier LLM | 64.14 | "inconsistent at managing 7 agents" |
| Conductor | trained 7B | **83.93** (LCB) | — |
| Conductor | untrained frontier (Table 11) | 67–70 (LCB) | over-relies on reputation |

The honest read is **not** "gitmoot's approach is wrong." It is: *a prompt-driven coordinator sits
at the weak end of this design space, and its biggest levers are (a) shrink the decision the LLM
has to make (role contracts, machine-checkable verdicts) and (b) give it an **outcome-feedback
signal** — because we cannot train a head.*

Two domain caveats gate everything below:
- Their "agents" are **stateless single-turn API calls** with a **cheap binary oracle reward**;
  gitmoot's are **long-running, stateful coding agents** mutating a real git checkout (branches,
  worktrees, PRs, locks) with **no clean reward** — CI/tests are the closest analog and are
  slow / partial / flaky.
- Their "Verifier" is *just another LLM*. gitmoot's verify phase can ground on **real tests**
  (strictly better) — but that also makes the ACCEPT signal noisier and more expensive than the
  papers' tidy ACCEPT/REVISE loop admits.

## Current orchestrate flow (for reference)

```mermaid
flowchart LR
  C[Coordinator job] -->|delegations DAG| D1[child: plan]
  C -->|delegations DAG| D2[child: implement]
  C -->|delegations DAG| D3[child: review]
  D1 --> K{synthesis_rule gate<br/>vote / quorum K}
  D2 --> K
  D3 --> K
  K -->|satisfied| CONT[Continuation job<br/>re-invokes coordinator]
  K -->|not satisfied| BLOCK[parent blocks]
  CONT -.bounded by.-> B[depth 8 · total jobs 64<br/>width 16 · 2h wall-clock<br/>loop detection]
```

## Insight catalog

Effort is "rough cost *if* pursued." "Builds on" names the existing primitive each idea layers on.

| # | Insight | Builds on (gitmoot primitive) | Paper | Effort | Tier |
|---|---|---|---|---|---|
| 1 | First-class **Thinker/Worker/Verifier** roles on `phase`, with a machine-checkable Verifier **ACCEPT/REVISE** verdict | `Delegation.Phase` (#369) + `decision` vocab | both | medium | 1 |
| 2 | **Verifier-gated revise loop** (implement→verify until ACCEPT or budget) | continuation job + termination bounds | TRINITY | low (recipe) | 1 |
| 3 | **Outcome-history scoreboard** injected into the coordinator prompt | SQLite job results (`internal/db`) + SkillOpt | Conductor | medium | 1 |
| 4 | **Verification-gated synthesis** (a branch that *passes tests* beats vote count) | `synthesis_rule`/`quorum` (#370) + inline `artifact_body` (#368) | Conductor | medium | 1 |
| 5 | **Cheap-Worker / strong-Verifier** model heterogeneity | per-delegation `model` | TRINITY | low (docs) | 2 |
| 6 | **Complementarity-driven pool** (codex + claude + kimi, not 3× the same) | council recipe | TRINITY | medium (guidance) | 2 |
| 7 | **Difficulty-adaptive fan-out** (assess complexity → size the council; 0 delegations for trivial) | coordinator prompt + bounds | Conductor | low | 2 |
| 8 | Per-edge **`context_from` visibility** (access-list) | generalizes inline `artifact_body` (#368) | Conductor | medium | 3 (marginal*) |
| 9 | First-class continuation **`decision: {accept, revise, escalate}`** | continuation + bounds | Conductor | low | 3 (marginal*) |
| 10 | **OOD exemplars** in coordinator prompt templates | SkillOpt / agent-templates | Conductor | low | 3 |
| 11 | **Per-Question-Best oracle** as orchestrate's north-star metric | eval lens over job history | TRINITY | low | framing |

\* Tier 3 items are flagged *by the papers' own ablations* as small / task-specific (see "What does not transfer").

## Tier 1 — the coherent, prompt-compatible thread

```mermaid
flowchart TD
  C[Coordinator continuation] --> ASSESS{assess complexity · #7}
  ASSESS -->|trivial| DONE0[do it · 0 delegations]
  ASSESS -->|complex| T[Thinker<br/>phase=thinker]
  T --> W[Worker<br/>phase=worker · cheap model #5]
  W --> V[Verifier<br/>phase=verifier · strong model<br/>decision: approved / changes_requested]
  V -->|quorum of approved >= K · #4| ACCEPT[terminate · synthesize<br/>prefer branch that PASSES tests]
  V -->|changes_requested| REVISE[bounded revise generation<br/>REVISE diagnosis inlined via artifact_body #4]
  REVISE --> W
  REVISE -.bounded by.-> B[depth 8 · total 64 · 2h<br/>loop detection]
```

**#1 — Role contracts + machine-checkable Verifier verdict (medium).**
Today `phase` is pure pass-through metadata with no contract. Define a small recognized role
vocabulary (`thinker` / `worker` / `verifier`) that, when set, injects a role-specific preamble
into the child (mirroring TRINITY's per-role "message processing module") **and requires verifier
children to return a normalized `decision`** — mapping `ACCEPT → approved`, `REVISE →
changes_requested` (both already in the contract). Then the continuation has the exact gate signal
it needs, and `quorum`/`vote` becomes "count verifier **ACCEPT**s" rather than counting arbitrary
child decisions. This is both papers' strongest shared structural idea and slots onto `phase` +
`synthesis_rule` directly.

**#2 — Verifier-gated revise loop (low, recipe only).**
The continuation job already re-invokes the coordinator after children finish — that *is* a turn
loop. Encode a coordinator pattern: on `changes_requested`, fan out a **bounded** revise→re-verify
generation; on `approved`, terminate. Our existing bounds (`MaxDelegationTotalJobs=64`,
`MaxDelegationDepth=8`, 2h wall-clock, loop detection) *are* the "fixed turn budget" backstop. No
engine change — document it as a recipe. *Caveat:* the ACCEPT signal is noisier/more expensive for
us than for them; ground it on real tests where possible.

**#3 — Outcome-history scoreboard (medium).**
The direct answer to "prompt coordinators over-rely on reputation." Don't train a head — persist a
lightweight repo/task-scoped scoreboard in SQLite (which runtime's branch passed CI / was accepted,
per task-type) and **inject it into the coordinator prompt as observed performance**, so runtime
selection is grounded in outcomes, not priors. The Conductor paper is explicit that anonymizing
agents *without* a capability table just makes selection random — so the scoreboard *is* the fix.
Substrate exists: job results already persist as JSON payload in SQLite (`internal/db/store.go`)
and SkillOpt (`internal/skillopt/`) already learns from history.

```mermaid
flowchart LR
  J[Job results · SQLite<br/>internal/db/store.go] --> S[Outcome scoreboard<br/>per repo · task-type · runtime]
  S -->|injected as observed performance| CP[Coordinator prompt]
  CP --> SEL[runtime/model selection<br/>grounded in outcomes, not reputation]
  SEL --> NEWJOB[new delegation] --> J
```

**#4 — Verification-gated synthesis (medium).**
`synthesis_rule` is a pure count today — exactly the MoA failure mode where one *verifiably-correct*
branch should beat several confidently-wrong ones. When children produce code, prefer the branch
that **passes tests/build**; let votes only break ties among passing branches. Pairs with inlining
each verifier's REVISE diagnosis (#368) so the revise-generation prompt names concrete defects, not
a bare "quorum not met."

## Tier 2 — cheap, doc/recipe-level

- **#5 Cheap-Worker / strong-Verifier (low).** The contract *already* supports per-delegation and
  per-ephemeral `model` (RESULT_CONTRACT even names "a cheap gate plus a strong verifier"). Make it
  a recommended, discoverable pattern: Worker cheap/fast, Verifier strong, Thinker reasoning.
- **#6 Complementarity-driven pool (medium guidance).** TRINITY shows pool *diversity*
  (optimized in error-space) is load-bearing — "3× strong but similar" leaves the oracle ceiling on
  the table. Practical version: bias councils toward one codex + one claude + one kimi. *Caveat:*
  the paper's exhaustive subset selection assumes you can cheaply run every agent on every task — we
  can't, so this is **guidance, not a live selection procedure.**
- **#7 Difficulty-adaptive fan-out (low).** Conductor's strong results come from *fewer, targeted*
  calls. Add an "assess complexity → size the council" step: one-line fix → 0 delegations; reserve
  the full plan→implement→review council for hard tasks. Directly relevant to our
  total-jobs/wall-clock bounds.

## Tier 3 — the papers' own ablations call these marginal

- **#8 Per-edge `context_from` visibility (medium).** Generalizes inline `artifact_body` (#368) to
  per-edge. **But Conductor's ablation (B.8) found the fine-grained topology gave *no significant*
  improvement at 7B** — they kept simple "all vs none." Adopt the simple version first; treat
  per-edge as likely over-engineering for now.
- **#9 `decision: {accept, revise, escalate}` continuation outcome (low).** Same spirit as #2;
  paper notes recursion gains were small, task-specific (+2.2pts BigCodeBench), OOD-only, ~2× cost →
  argues for **bounded, selective** re-delegation, not always-on.
- **#10 OOD exemplars in templates (low).** Teach plan *shape* with an example from a *different*
  task family to avoid copy-pasting a stale strategy. A cheap additive heuristic for
  SkillOpt / agent-templates.

## North-star metric (framing, free)

**#11 — Per-Question-Best oracle.** gitmoot orchestrate *is* macro-level fusion of black-box coding
agents that can never be weight-merged — the papers are the strongest external endorsement of the
thesis. As an eval lens: log per-agent **solo** outcomes on a task corpus and compute
**best-single** vs **any-agent-correct (the ceiling)** vs **orchestrate result**. Gap to ceiling =
headroom; gap over best-single = value-add. *Caveat:* building the full performance matrix means
running every agent on every task — expensive/side-effecting for coding — so treat it as a
**sampled evaluation north-star, not a live selection procedure.**

## What does NOT transfer (non-goals)

- **The training machinery.** sep-CMA-ES / GRPO, the 0.6B SLM hidden-state features, the ~10K-param
  head, block-ε-separability theory (Props 1–2), reward shaping. gitmoot's coordinator is a frozen
  frontier model driven by prompts + the `gitmoot_result` contract — there is no training loop.
- **The headline numbers.** 86.2% LCB / 83.93 belong to the *trained* coordinators; the
  *untrained-frontier* baselines (gitmoot's regime) were materially worse. Don't quote them as
  achievable here.
- **The cheap binary oracle reward.** Their tasks are single-shot, auto-verifiable, short-horizon
  (≤5 turns, 4096-token caps). gitmoot's coding work is long-horizon and side-effecting with no
  clean reward — verifier-gated iteration is sound *in spirit* but the ACCEPT signal is noisier and
  costlier.
- **Per-edge topology control (#8)** — ablation-marginal at current model sizes.
- **The "coordination is token-efficient" claim.** Their budgets were tiny *to make RL tractable*;
  gitmoot fan-out spawns full agent sessions (~N× the cost of one agent). The cheap-worker/
  strong-verifier pattern (#5) is the right lever; the blanket efficiency claim does not carry over.

## If we ever pursue this

The coherent, highest-value thread is **#1 → #2 → #4 → #3**: the first three are recipe/contract
work that builds straight on the `phase` / `quorum` / `inline_artifact_body` primitives shipped in
#369 / #370 / #368; **#3 (outcome scoreboard)** is the one genuinely new subsystem and the real
counter to "prompt coordinators lean on reputation." Everything in Tier 3 the papers themselves
flag as marginal. Recorded here for whoever picks it up — not scheduled.

**Comment by jerryfane** [comment]
Closing as **not a work item**: the body self-describes as a note or reference rather than a deliverable, and nothing in it is scheduled.

**Evidence (LIVE):** Body self-disclaims: "a research reference, not a build commitment… Mine it later; nothing here is scheduled" — no obligation encoded, 90d idle, 0 comments.



---
*From the 2026-09-21 backlog triage: all 439 unparented open issues read at body level, claims re-measured against the checkout at `c88c2614` and the live store. Full evidence set in `/root/fleet-tools/state/backlog-triage-2026-09-21/`.*

**Comment by jerryfane** [comment]
Closing as **not a work item**: the body self-describes as a note or reference rather than a deliverable, and nothing in it is scheduled.

**Evidence (LIVE):** Body self-disclaims: "a research reference, not a build commitment… Mine it later; nothing here is scheduled" — no obligation encoded, 90d idle, 0 comments.



---
*From the 2026-09-21 backlog triage: all 439 unparented open issues read at body level, claims re-measured against the checkout at `c88c2614` and the live store. Full evidence set in `/root/fleet-tools/state/backlog-triage-2026-09-21/`.*
